### PR TITLE
Move database config to flytestdlib

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
We have the same database section in flyteadmin and the datacatalog. Therefore, move them to stdlib and avoid errors when launching Flyte cluster by using single binary.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2266

## Follow-up issue
_NA_

### CODE REVIEW

File: config/database/config.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package database
```
Comment (EngHabu) [inline config/database/config.go:1]: I think we should move this to a top level package `database`

File: config/database/config.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/flyteorg/flytestdlib/config"
+	"gorm.io/gorm/logger"
+)
+
+const (
+	database = "database"
+	postgres = "postgres"
+)
+
+var Config = config.MustRegisterSection(database, &DbConfig{
```
Comment (EngHabu) [inline config/database/config.go:15]:
```suggestion
var configSection = config.MustRegisterSection(database, &DbConfig{
```

Comment (EngHabu) [inline config/database/config.go:15]: Let's add `//go:generate pflags.......` here as well

File: config/database/config.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	User   string `json:"username" pflag:",The database user who is connecting to the server."`
+	// Either Password or PasswordPath must be set.
+	Password     string `json:"password" pflag:",The database password."`
+	PasswordPath string `json:"passwordPath" pflag:",Points to the file containing the database password."`
+	ExtraOptions string `json:"options" pflag:",See http://gorm.io/docs/connecting_to_the_database.html for available options passed, in addition to the above."`
+	Debug        bool   `json:"debug" pflag:" Whether or not to start the database connection with debug mode enabled."`
+}
+
+func GetConfig() *DbConfig {
+	return Config.GetConfig().(*DbConfig)
```
Comment (EngHabu) [inline config/database/config.go:66]: I think we should also move the merging logic for the deprecated fields here...

**Review by EngHabu — COMMENTED** [review]
Awesome! thank you for taking care of this

File: config/database/config.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	DeprecatedExtraOptions string `json:"options" pflag:",deprecated"`
+	DeprecatedDebug        bool   `json:"debug" pflag:",deprecated"`
+
+	EnableForeignKeyConstraintWhenMigrating bool            `json:"enableForeignKeyConstraintWhenMigrating" pflag:",Whether to enable gorm foreign keys when migrating the db"`
+	MaxIdleConnections                      int             `json:"maxIdleConnections" pflag:",maxIdleConnections sets the maximum number of connections in the idle connection pool."`
+	MaxOpenConnections                      int             `json:"maxOpenConnections" pflag:",maxOpenConnections sets the maximum number of open connections to the database."`
+	ConnMaxLifeTime                         config.Duration `json:"connMaxLifeTime" pflag:",sets the maximum amount of time a connection may be reused"`
+	LogLevel                                logger.LogLevel `json:"log_level" pflag:"-,"`
+	PostgresConfig                          *PostgresConfig `json:"postgres,omitempty"`
+	SQLiteConfig                            *SQLiteConfig   `json:"sqlite,omitempty"`
```
Comment (EngHabu) [inline config/database/config.go:44]:
```suggestion
	Postgres                          *PostgresConfig `json:"postgres,omitempty"`
	SQLite                            *SQLiteConfig   `json:"sqlite,omitempty"`
```

File: config/database/config.go (hunk 2 above)
Comment (EngHabu) [inline config/database/config.go:15]: We shouldn't use the deprecated fields in here... let's use the new fields only...

Comment (pingsutw) [reply config/database/config.go]: Updated it

File: database/config.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+// entities (e.g. workflows, tasks, launch plans...)
+type DbConfig struct {
+	DeprecatedHost         string `json:"host" pflag:",deprecated"`
+	DeprecatedPort         int    `json:"port" pflag:",deprecated"`
+	DeprecatedDbName       string `json:"dbname" pflag:",deprecated"`
+	DeprecatedUser         string `json:"username" pflag:",deprecated"`
+	DeprecatedPassword     string `json:"password" pflag:",deprecated"`
+	DeprecatedPasswordPath string `json:"passwordPath" pflag:",deprecated"`
+	DeprecatedExtraOptions string `json:"options" pflag:",deprecated"`
+	DeprecatedDebug        bool   `json:"debug" pflag:",deprecated"`
```
Comment (EngHabu) [inline database/config.go:41]:
```suggestion
        // deprecated: Please use Postgres.Host or SQLLite.Host instead
	DeprecatedHost         string `json:"host" pflag:"-,deprecated"`
	
	// deprecated:... <same for all properties>
	DeprecatedPort         int    `json:"port" pflag:"-,deprecated"`
	DeprecatedDbName       string `json:"dbname" pflag:"-,deprecated"`
	DeprecatedUser         string `json:"username" pflag:"-,deprecated"`
	DeprecatedPassword     string `json:"password" pflag:"-,deprecated"`
	DeprecatedPasswordPath string `json:"passwordPath" pflag:"-,deprecated"`
	DeprecatedExtraOptions string `json:"options" pflag:"-,deprecated"`
	DeprecatedDebug        bool   `json:"debug" pflag:"-,deprecated"`
```

Comment (pingsutw) [reply database/config.go]: updated it, I didn't add SQLLite.Host in the comment because we only have one property `File` in `SQLiteConfig`

**Review by kumare3 — APPROVED** [review]

File: database/config.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	Password     string `json:"password" pflag:",The database password."`
+	PasswordPath string `json:"passwordPath" pflag:",Points to the file containing the database password."`
+	ExtraOptions string `json:"options" pflag:",See http://gorm.io/docs/connecting_to_the_database.html for available options passed, in addition to the above."`
+	Debug        bool   `json:"debug" pflag:" Whether or not to start the database connection with debug mode enabled."`
+}
+
+func GetConfig() *DbConfig {
+	databaseConfig := configSection.GetConfig().(*DbConfig)
+	if databaseConfig.Postgres == nil && databaseConfig.SQLite == nil {
+		databaseConfig.Postgres = &PostgresConfig{}
```
Comment (EngHabu) [inline database/config.go:81]: This only initializes Postgres but checks for both... it can lead to nil ref if SQLite and DeprecatedHost are set...

File: database/config.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	// deprecated: Please use Postgres.Debug
+	DeprecatedDebug bool `json:"debug" pflag:",deprecated"`
+
+	EnableForeignKeyConstraintWhenMigrating bool            `json:"enableForeignKeyConstraintWhenMigrating" pflag:",Whether to enable gorm foreign keys when migrating the db"`
+	MaxIdleConnections                      int             `json:"maxIdleConnections" pflag:",maxIdleConnections sets the maximum number of connections in the idle connection pool."`
+	MaxOpenConnections                      int             `json:"maxOpenConnections" pflag:",maxOpenConnections sets the maximum number of open connections to the database."`
+	ConnMaxLifeTime                         config.Duration `json:"connMaxLifeTime" pflag:",sets the maximum amount of time a connection may be reused"`
+	LogLevel                                logger.LogLevel `json:"log_level" pflag:"-,"`
+	Postgres                                *PostgresConfig `json:"postgres,omitempty"`
+	SQLite                                  *SQLiteConfig   `json:"sqlite,omitempty"`
```
Comment (EngHabu) [inline database/config.go:57]: I would say let's keep these non-pointers... they lead to nasty nil refs...
```suggestion
	Postgres                               PostgresConfig `json:"postgres,omitempty"`
	SQLite                                  SQLiteConfig   `json:"sqlite,omitempty"`
```

File: database/config.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// SQLiteConfig can be used to configure
+type SQLiteConfig struct {
```
Comment (EngHabu) [inline database/config.go:61]: Let's add this method to both:
```golang
var emptySQLiteCfg = SQLiteConfig{}
func (s SQLiteConfig) IsEmpty() bool {
  return s == emptySQLiteCfg
}
```

File: config/database/config.go (hunk 2 above)
Comment (EngHabu) [reply config/database/config.go]: Let's do this too... and run generate..

File: database/config.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	// deprecated: Please use Postgres.User
+	DeprecatedUser string `json:"username" pflag:",deprecated"`
+	// deprecated: Please use Postgres.Password
+	DeprecatedPassword string `json:"password" pflag:",deprecated"`
+	// deprecated: Please use Postgres.PasswordPath
+	DeprecatedPasswordPath string `json:"passwordPath" pflag:",deprecated"`
+	// deprecated: Please use Postgres.ExtraOptions
+	DeprecatedExtraOptions string `json:"options" pflag:",deprecated"`
+	// deprecated: Please use Postgres.Debug
+	DeprecatedDebug bool `json:"debug" pflag:",deprecated"`
```
Comment (EngHabu) [inline database/config.go:49]:
```suggestion
	DeprecatedHost string `json:"host" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.Port
	DeprecatedPort int `json:"port" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.DbName
	DeprecatedDbName string `json:"dbname" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.User
	DeprecatedUser string `json:"username" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.Password
	DeprecatedPassword string `json:"password" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.PasswordPath
	DeprecatedPasswordPath string `json:"passwordPath" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.ExtraOptions
	DeprecatedExtraOptions string `json:"options" pflag:"-,deprecated"`
	// deprecated: Please use Postgres.Debug
	DeprecatedDebug bool `json:"debug" pflag:"-,deprecated"`
```

File: database/dbconfig_flags.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+func (cfg DbConfig) GetPFlagSet(prefix string) *pflag.FlagSet {
+	cmdFlags := pflag.NewFlagSet("DbConfig", pflag.ExitOnError)
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "host"), defaultConfig.DeprecatedHost, "deprecated")
+	cmdFlags.Int(fmt.Sprintf("%v%v", prefix, "port"), defaultConfig.DeprecatedPort, "deprecated")
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "dbname"), defaultConfig.DeprecatedDbName, "deprecated")
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "username"), defaultConfig.DeprecatedUser, "deprecated")
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "password"), defaultConfig.DeprecatedPassword, "deprecated")
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "passwordPath"), defaultConfig.DeprecatedPasswordPath, "deprecated")
+	cmdFlags.String(fmt.Sprintf("%v%v", prefix, "options"), defaultConfig.DeprecatedExtraOptions, "deprecated")
+	cmdFlags.Bool(fmt.Sprintf("%v%v", prefix, "debug"), defaultConfig.DeprecatedDebug, "deprecated")
```
Comment (EngHabu) [inline database/dbconfig_flags.go:60]: These are not good... let's add the change above and regenerate to avoid generating these lines

**Review by EngHabu — COMMENTED** [review]
One nit... Thank you!

**Review by EngHabu — APPROVED** [review]
Thank you!
